File: ios/Sources/AppDiagLog/Crypto/SessionEnvelope.swift
import Foundation

/// On-disk session envelope format
struct SessionEnvelope: Codable, Sendable {
    let version: Int
    let sessionId: String
    let createdAt: String
    let sealedAt: String?
    let eventCount: Int
    let sessionTag: String?
    let deviceMetadata: [String: String]
    let encryption: Encryption
    let payload: String

    struct Encryption: Codable, Sendable {
        let algorithm: String      // "AES-256-GCM" | "AES-128-GCM" | "ChaCha20-Poly1305"
        let nonce: String
        let kekAlgorithm: String   // "ML-KEM-768" | "ML-KEM-512" | "RSA-OAEP-3072" | "ECDH-P256+HKDF"
        let keyId: String
        let kemCiphertext: String  // base64 (empty string for RSA-OAEP)
        let wrappedDek: String     // base64
        /// Algorithm-specific extras. Encoded only when non-empty.
        let kekParams: [String: String]?

        enum CodingKeys: String, CodingKey {
            case algorithm, nonce
            case kekAlgorithm = "kek_algorithm"
            case keyId = "key_id"
            case kemCiphertext = "kem_ciphertext"
            case wrappedDek = "wrapped_dek"
            case kekParams = "kek_params"
        }
    }

    enum CodingKeys: String, CodingKey {
        case version, encryption, payload
        case sessionId = "session_id"
        case createdAt = "created_at"
        case sealedAt = "sealed_at"
        case eventCount = "event_count"
        case sessionTag = "session_tag"
        case deviceMetadata = "device_metadata"
    }
}

File: backend/src/main/kotlin/com/appdiaglog/server/model/SessionEnvelope.kt
package com.appdiaglog.server.model

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

/**
 * Wire-compatible mirror of the SDK's `session_*.enc` envelope.
 *
 * Keys are snake_case to match what iOS (Codable with CodingKeys) write. Adding a field requires
 * updating all three layers in lockstep.
 *
 * `payload` and the inner crypto fields are base64. We keep them as strings here
 * so Jackson never has to decode binary while parsing — decoding happens in the
 * decryption pipeline where errors map cleanly to "bad ciphertext" responses.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
data class SessionEnvelope(
    val version: Int,
    @JsonProperty("session_id") val sessionId: String,
    @JsonProperty("created_at") val createdAt: String,
    @JsonProperty("sealed_at") val sealedAt: String? = null,
    @JsonProperty("event_count") val eventCount: Int,
    @JsonProperty("session_tag") val sessionTag: String? = null,
    @JsonProperty("device_metadata") val deviceMetadata: Map<String, String> = emptyMap(),
    val encryption: Encryption,
    val payload: String,
) {
    @JsonInclude(JsonInclude.Include.NON_NULL)
    data class Encryption(
        val algorithm: String,
        val nonce: String,
        @JsonProperty("kek_algorithm") val kekAlgorithm: String,
        @JsonProperty("key_id") val keyId: String,
        @JsonProperty("kem_ciphertext") val kemCiphertext: String,
        @JsonProperty("wrapped_dek") val wrappedDek: String,
        /** Optional algorithm-specific parameters. Pre-agility envelopes omit it. */
        @JsonProperty("kek_params") val kekParams: Map<String, String>? = null,
    )
}
